src/software/ai/evaluation/q_learning/attacker_mdp_feature_extractor.cpp

```diff
@@ -0,0 +1,82 @@
+#include "software/ai/evaluation/q_learning/attacker_mdp_feature_extractor.h"
+
+#include <math.h>
+
+#include "shared/constants.h"
+#include "software/ai/evaluation/possession.h"
+#include "software/math/math_functions.h"
+
+AttackerMdpFeatureExtractor::AttackerMdpFeatureExtractor()
+    : FeatureExtractor({{"ball_x_position", ballXPositionFeature},
+                        {"best_pass_rating", bestPassRatingFeature},
+                        {"best_shot_open_angle", bestShotOpenAngleFeature},
+                        {"nearby_enemy_threats", nearbyEnemyThreatsFeature},
+                        {"num_friendly_robots", numFriendlyRobotsFeature},
+                        {"num_enemy_robots", numEnemyRobotsFeature}})
+{
+}
+
+double AttackerMdpFeatureExtractor::ballXPositionFeature(const AttackerMdpState& state)
+{
+    const double x_pos            = state.world_ptr->ball().position().x();
+    const double field_x_length   = state.world_ptr->field().totalXLength();
+    const double normalized_x_pos = std::max(x_pos, 0.0) / (field_x_length / 2);
+    return normalized_x_pos;
+}
+
+double AttackerMdpFeatureExtractor::bestPassRatingFeature(const AttackerMdpState& state)
+{
+    return state.strategy->getBestPass().rating;
+}
+
+double AttackerMdpFeatureExtractor::bestShotOpenAngleFeature(
+    const AttackerMdpState& state)
+{
+    std::optional<Robot> attacker = getRobotWithEffectiveBallPossession(
+        state.world_ptr->friendlyTeam(), state.world_ptr->ball(),
+        state.world_ptr->field());
+
+    if (attacker)
+    {
+        std::optional<Shot> best_shot = state.strategy->getBestShot(*attacker);
+        if (best_shot)
+        {
+            return best_shot->getOpenAngle().toDegrees() * SHOT_OPEN_ANGLE_FACTOR;
+        }
```

Comment:
If I'm not wrong, every feature except this one is normalized between [0, 1]. How come this isn't as well?

Reply:
I guess it gets normalized by the `FeatureExtractor` base class